pkg/readychecker/types/built_in.go

```diff
@@ -62,6 +77,49 @@ func isDeploymentReady(obj runtime.Object) (isReady, retriableError bool, e erro
 		deployment.Status.UpdatedReplicas == replicas, false, nil
 }
 
+func isHorizontalPodAutoscalerReady(obj runtime.Object) (isReady, retriableError bool, e error) {
+	var hpa autoscaling_v2b1.HorizontalPodAutoscaler
+	if err := util.ConvertType(autoscalingV2B1, obj, &hpa); err != nil {
+		return false, false, err
+	}
+
+	// For the HPA to be ready, AbleToScale and ScalingActive conditions should exist and be true
+	// If ScalingActive is false, it could still be coming up *or* it could have failed (e.g. because it can't access metrics)
+	// Can't distinguish this based on the reason, so using a timeout
+	foundAbleToScale := false
+	foundScalingActive := false
+	for _, cond := range hpa.Status.Conditions {
+		switch cond.Type {
+		case hpaAbleToScaleCondition:
+			switch cond.Status {
+			case core_v1.ConditionTrue:
+				foundAbleToScale = true
+			case core_v1.ConditionFalse:
+				// AbleToScale should not be false if the HPA is working, this is a failure
+				return false, false, errors.Errorf("%s: %s", cond.Reason, cond.Message)
```

Comment:
@ash2k is that correct to return a non-retriable error in that case?
```
	// AbleToScale indicates a lack of transient issues which prevent scaling from occurring,
	// such as being in a backoff window, or being unable to access/update the target scale.
	AbleToScale HorizontalPodAutoscalerConditionType = "AbleToScale"
```
In case of backoff this error is recoverable, and updating the target scale could succeed after retrying...
It doesn't affect the lifecycle of the bundle AFAICT, it's mostly about reporting `InProgress: True` with error `Reason: RetriableError` vs `InProgress: False` and error `Reason: TerminalError`.

Reply:
Yes and no. We need to refactor error propagation in Smith. You are correct, that it influences the reason smith reports. But it also retries to re-process things in the error is marked as retriable, which will not help in this case (but is also not very harmful, only extra logs). So perhaps in this case the error should be marked as retriable to be future compatible with what we will hopefully do to surface such conditions. See https://github.com/atlassian/smith/issues/276

Reply:
Yep, I also had a thought conflict between correct status and Smith indefinitely stuck in retry loop.

Reply:
I have made these retriable.